### ISSUE TITLE
🐛 [Bug Report] Great Stamina not affecting pet stamina and heath correctly

### ISSUE THREAD

**Issue by MilkmanMHOC** [cmangos/issues#3817]
### Bug Details

Training your pet with any stamina under rank 6 does not apply correctly?  While stamina is increased health is not affected by the increase.  Here is a specific example.  I have a lvl 15 boar.  By default he has a stamina of 51 and 392 health.  I train him in great stamina 1.  He now has 54 stamina but still has 392 health.  I then train the same boar to rank 2, and he now has 56 stamina, however his health is still 392.    This issue continues on all the various pets i have tested, and stamina doesn't seem to get affected until much higher levels.  

### Steps to Reproduce

1. Train any hunter pet with great stamina
2. note health does not increase
3.
4.

### Expected behavior

As you increase your pets stamina your pets health should increase appropriately.  

### Core SHA1 Commit Hash

e593dbcde632ffc4ba04e82ffc01a51695d578a8

### Database SHA1 Commit Hash

347e3c062be5c7026eb925f2154af81555b3865d

### Operating System

win7

### Client Version

1.12.1 (Classic)

**Comment by MilkmanMHOC** [comment]
Of note, i did find this same issues reported long ago.. and dismissed as not reproducible? I have seen this error for some time, but i had not updated my server until recently so wasn't sure it was still an issue.  Unfortunately it is.  I do not know if the details in the old ticket are still valid or not?  https://github.com/cmangos/issues/issues/1896

**Comment by MilkmanMHOC** [comment]
one more thought, i wonder if endurance training talent could be affecting this?  I do have 5 skills of that on all my hunters i was testing with.

**Comment by Pip1987** [comment]
So I can confirm Ranks 1-6 Greater Stamina are increasing pet stam but not HP. Ranks 7-10 are working as intended. I do not have any points in Endurance Training.

**Comment by Pip1987** [comment]
I think the culprit of the issue is the file StatSystem.cpp line 859 ``value += GetModifierValue(unitMod, TOTAL_VALUE) + std::max((stamina - 20) * 10 + 20, 0.f);`` Ranks 1-6 increase stamina by less than 20, resulting in 0 thus the stamina modifier is rendered useless from rank 6 and below is useless, and possibly the incorrect increase for ranks 7-10.

![image](https://github.com/user-attachments/assets/1adae53b-ba16-4c65-8cb9-eeab00165e42)
![image](https://github.com/user-attachments/assets/01187944-c6e9-467e-85b8-d816648ddb07)
![image](https://github.com/user-attachments/assets/7e4c0d84-face-4a85-bb36-1312bfb01849)

**Comment by insunaa** [comment]
I see nowhere where the increase is set to 0 in StatSystem

**Comment by killerwife** [comment]
First 20 points of stamina increase on a player do nothing. I modelled the functions for pet the same way but that seems to be wrong. (due to weird base stamina interaction)

This is definitely a bug, leave it to me.

**Comment by Pip1987** [comment]
> I see nowhere where the increase is set to 0 in StatSystem

In the formula ``value += GetModifierValue(unitMod, TOTAL_VALUE) + std::max((stamina - 20) * 10 + 20, 0.f);`` stamina is equal to the difference between the creature's base stamina and current stamina. So when stamina is 3 (the increase from Rank 1) then std::Max( (3-20) * 10) +20, 0); is zero because 3-20 would be -17, multiply that by 10 would be -170 + 20 = -150.

**Comment by insunaa** [comment]
yeah, I thought you meant no increase at all, even unitmods. I would think that Greater Stamina would be an Aura or the like so fall into UnitMods and not into base stamina

**Comment by Pip1987** [comment]
> yeah, I thought you meant no increase at all, even unitmods. I would think that Greater Stamina would be an Aura or the like so fall into UnitMods and not into base stamina

yeah I thought the same, I'm leveling a hunter in classic right now to see how it looks, I thought it was strange that the stamina is just raised with no indication of it not being base.
